cpp/ql/src/semmle/code/cpp/commons/Exclusions.qll

```diff
@@ -94,11 +94,22 @@ predicate functionContainsPreprocCode(Function f) {
  * ```
  */
 predicate isFromMacroDefinition(Element e) {
-  exists(MacroInvocation mi |
+  exists(MacroInvocation mi, Location eLocation, Location miLocation |
     // e is in mi
     mi.getAnExpandedElement() = e and
     // and e was apparently not passed in as a macro parameter
-    e.getLocation().getStartLine() = mi.getLocation().getStartLine() and
-    e.getLocation().getStartColumn() = mi.getLocation().getStartColumn()
+    eLocation = e.getLocation() and
+    miLocation = mi.getLocation() and
+    nonBindingIntEquality(eLocation.getStartLine(), miLocation.getStartLine()) and
+    nonBindingIntEquality(eLocation.getStartColumn(), miLocation.getStartColumn())
   )
 }
+
+/**
+ * Holds if `x = y` but gets compiled to a filter instead of a join. This can
+ * be used to avoid bad join orders where integers are joined too early.
+ */
+bindingset[x, y]
+private predicate nonBindingIntEquality(int x, int y) {
+  x >= y and y >= x
```

Comment:
This is ... interesting.

Reply:
It's definitely a hack, but I think I can tolerate it given that it's clearly named and documented.

Reply:
It's a trick that's used a lot in the shared data flow library. The predicate there is called `unbind`, and some of us are hoping that such a predicate will one day be built into the language.